go/vt/vitessdriver/convert.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2017 GitHub Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreedto in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vitessdriver
+
+import (
+	"database/sql/driver"
+	"fmt"
+	"time"
+
+	"github.com/youtube/vitess/go/sqltypes"
+	querypb "github.com/youtube/vitess/go/vt/proto/query"
+)
+
+type converter struct {
+	datetime bool
```

Comment:
For those following at home: why would we even consider adding a flag to enable/disable time conversion?

In theory, since the driver has never supported `time.Time` conversion, this should _not_ be a compatibility-breaking change: before this PR, attempting to `Scan` a `time.Time` would return an error; after this PR, if the conversion is enabled, the `Scan` will succeed.

So where are we breaking backwards compatibility? When users were scanning `DATETIME` or `DATE` values directly into `[]byte` or `string` variables. Before this PR, that was the only way to scan a `DATETIME` column; you'd need to pass in a `[]byte` to the `Scan` method and the driver would return the raw value, straight from MySQL, represented as `"2006-01-02 15:04:05.999999"`.

However, once this PR is merged, if `time.Time` conversion is enabled by default, or always enabled, any `Scan` callsites that scan a `DATETIME` column into a `[]byte` will receive a different output! Since the conversion happens at the driver level (like all driver conversions), `database/sql` will never get to see the raw MySQL form for datetimes.

Instead, `database/sql` will receive a `time.Time` value, and will attempt to coerce it into a `[]byte` or `string`:

https://github.com/golang/go/blob/38ad33088c3bee63e8d53e7aff4d6610d82ca84c/src/database/sql/convert.go#L265-L276

The `database/sql` coercion rules use `time.RFC3339Nano` to format the `Time`. Hence, the result of the `Scan` will _not_ match the previous behaviour of the driver (`2006-01-02T15:04:05.999999999Z07:00` vs `2006-01-02 15:04:05.999999`).

That is the breaking change, and that's the reason why this PR doesn't enable `time.Time` conversion by default.

Hope this explanation is useful to people reviewing the code. For our particular use case, we intend to always keep `time.Time` conversion enabled, and we don't currently use the driver in production yet, so I'm leaning towards @sougou's suggestion.

I'm going to update the PR and we can wait to see who complains. :)